components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/basicauth/BasicAuthCredentialValidator.java

```diff
@@ -0,0 +1,478 @@
+import org.wso2.carbon.apimgt.gateway.internal.ServiceReferenceHolder;
+import org.wso2.carbon.apimgt.impl.APIConstants;
+import org.wso2.carbon.apimgt.impl.APIManagerConfiguration;
+import org.wso2.carbon.apimgt.impl.utils.APIUtil;
+import org.wso2.carbon.authenticator.stub.AuthenticationAdminStub;
+import org.wso2.carbon.authenticator.stub.LoginAuthenticationExceptionException;
+import org.wso2.carbon.base.ServerConfiguration;
+import org.wso2.carbon.um.ws.api.stub.RemoteUserStoreManagerServiceStub;
+import org.wso2.carbon.um.ws.api.stub.RemoteUserStoreManagerServiceUserStoreExceptionException;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.user.api.UserStoreManager;
+import org.wso2.carbon.utils.CarbonUtils;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import javax.cache.Cache;
+import javax.cache.Caching;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.rmi.RemoteException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Map;
+
+/**
+ * This class will validate the basic auth credentials.
+ */
+public class BasicAuthCredentialValidator {
+
+    private boolean gatewayKeyCacheEnabled;
+    private static boolean gatewayUsernameCacheInit = false;
+    private static boolean gatewayBasicAuthResourceCacheInit = false;
+
+    protected Log log = LogFactory.getLog(getClass());
+    private AuthenticationAdminStub authAdminStub;
+    private RemoteUserStoreManagerServiceStub remoteUserStoreManagerServiceStub;
+    private String host;
+
+    /**
+     * Initialize the validator.
+     */
+    BasicAuthCredentialValidator() {
+    }
+
+    /**
+     * Initialize the validator with the synapse environment.
+     *
+     * @param env the synapse environment
+     * @throws APISecurityException If an authentication failure or some other error occurs
+     */
+    BasicAuthCredentialValidator(SynapseEnvironment env) throws APISecurityException {
+        this.gatewayKeyCacheEnabled = isGatewayTokenCacheEnabled();
+        this.getGatewayUsernameCache();
+
+        ConfigurationContext configurationContext = ServiceReferenceHolder.getInstance().getAxis2ConfigurationContext();
+        APIManagerConfiguration config = ServiceReferenceHolder.getInstance().getAPIManagerConfiguration();
+        String url = config.getFirstProperty(APIConstants.API_KEY_VALIDATOR_URL);
+        if (url == null) {
+            throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                    "API key manager URL unspecified");
+        }
+
+        try {
+            authAdminStub = new AuthenticationAdminStub(configurationContext, url +
+                    "AuthenticationAdmin");
+        } catch (AxisFault axisFault) {
+            throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR, axisFault.getMessage());
```

Comment:
Exception has been swallowed leading to the original stack trace being lost. Pass entire exception object into wrapped exception

`catch (CaughtExecption e) {
      throw new ThrownException("Error detected", e);
}`

Reply:
Updated